oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java

```diff
@@ -521,6 +614,121 @@ private VersionGCStats gc(long maxRevisionAgeInMillis) throws IOException {
+         * commit related garbage.
+         * <p/>
+         * TODO: limit this to run only on a singleton instance, eg the cluster leader
+         * <p/>
+         * The "detail garbage" collector can be instructed to do a full repository scan
+         * - or incrementally based on where it last left off. When doing a full
+         * repository scan (but not limited to that), it executes in (small) batches
+         * followed by voluntary paused (aka throttling) to avoid excessive load on the
+         * system. The full repository scan does not have to finish particularly fast,
+         * it is okay that it takes a considerable amount of time.
+         *
+         * @param phases {@link GCPhases}
+         * @param headRevision the current head revision of node store
+         * @param rec {@link VersionGCRecommendations} to recommend GC operation
+         */
+        private void collectDetailedGarbage(final GCPhases phases, final RevisionVector headRevision, final VersionGCRecommendations rec)
+                throws IOException {
+            int docsTraversed = 0;
+            boolean foundDoc = true;
+            final long oldestModifiedMs = rec.scopeDetailedGC.fromMs;
+            final long toModified = rec.scopeDetailedGC.toMs;
+            long oldModifiedMs = oldestModifiedMs;
+            final String oldestModifiedDocId = rec.detailedGCId;
+            try (DetailedGC gc = new DetailedGC(headRevision, monitor, cancel)) {
+                long fromModified = oldestModifiedMs;
+                String fromId = ofNullable(oldestModifiedDocId).orElse(MIN_ID_VALUE);
+                NodeDocument lastDoc;
+                if (phases.start(GCPhase.DETAILED_GC)) {
+                    while (foundDoc && fromModified < toModified && docsTraversed < PROGRESS_BATCH_SIZE) {
+                        // set foundDoc to false to allow exiting the while loop
+                        foundDoc = false;
+                        lastDoc = null;
+                        Iterable<NodeDocument> itr = versionStore.getModifiedDocs(fromModified, toModified, 1000, fromId);
+                        try {
+                            for (NodeDocument doc : itr) {
+                                foundDoc = true;
+                                // continue with GC?
+                                if (cancel.get()) {
+                                    foundDoc = false; // to exit while loop as well
+                                    log.info("Received GC cancel call. Terminating the GC Operation.");
+                                    break;
+                                }
+                                docsTraversed++;
+                                if (docsTraversed % 100 == 0) {
+                                    monitor.info("Iterated through {} documents so far. {} had detail garbage",
+                                            docsTraversed, gc.getGarbageDocsCount());
```

Comment:
It looks like the two parameters here are not in sync : the `docsTraversed` is the total number of documents traversed in this entire G run - while `getGarbageDocsCount` is only covering 1 iteration (inner loop).